api/error_unexpected_response_test.go

```diff
@@ -0,0 +1,265 @@
+func testNotFoundHandler(b string) http.Handler { return testHandler(http.StatusNotFound, b) }
+
+// testNotFoundHandler creates a testHandler preconfigured with status code 200.
+func testOKHandler(b string) http.Handler { return testHandler(http.StatusOK, b) }
+
+// testHandler is a helper function that writes a Nomad-like server response
+// with the necessary headers to make the API client happy
+func testHandler(sc int, b string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		addMockHeaders(w.Header())
+		w.WriteHeader(sc)
+		w.Write([]byte(b))
+	})
+}
+
+// closingHandler is a handler that terminates the response body early in the
+// reading process
+func closingHandler(sc int, b string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		// We need a misbehaving reader to test network effects when collecting
+		// the http.Response data into a UnexpectedResponseError
+		er := iotest.TimeoutReader( // TimeoutReader throws an error on the second read
+			iotest.OneByteReader( // OneByteReader yields a byte at a time, causing multiple reads
+				strings.NewReader(mockNamespaceBody),
+			),
+		)
+
+		// We need to set content-length to the true value it _should_ be so the
+		// API-side reader knows it's a short read.
+		w.Header().Set("content-length", fmt.Sprint(len(mockNamespaceBody)))
+		addMockHeaders(w.Header())
+		w.WriteHeader(sc)
+
+		// Using io.Copy to send the data into w prevents golang from setting the
+		// content-length itself.
+		io.Copy(w, er)
+	})
+}
+
+// testLogRequestHandler wraps a http.Handler with a logger that writes to the
+// test log output
+func testLogRequestHandler(t *testing.T, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// call the original http.Handler wrapped in a httpsnoop
+		m := httpsnoop.CaptureMetrics(h, w, r)
+		ri := httpReqInfo{
+			uri:       r.URL.String(),
+			method:    r.Method,
+			ipaddr:    ipAddrFromRemoteAddr(r.RemoteAddr),
+			code:      m.Code,
+			duration:  m.Duration,
+			size:      m.Written,
+			userAgent: r.UserAgent(),
+		}
+		t.Logf(ri.String())
+	})
+}
+
+// httpReqInfo holds all the information used to log a request to the mock server
+type httpReqInfo struct {
+	method    string
+	uri       string
+	referer   string
+	ipaddr    string
+	code      int
+	size      int64
+	duration  time.Duration
+	userAgent string
+}
+
+func (i httpReqInfo) String() string {
+	return fmt.Sprintf(
+		"method=%q uri=%q referer=%q ipaddr=%q code=%d size=%d duration=%q userAgent=%q",
+		i.method, i.uri, i.referer, i.ipaddr, i.code, i.size, i.duration, i.userAgent,
+	)
+}
+
+// ipAddrFromRemoteAddr removes the port from the address:port in remote addr
+func ipAddrFromRemoteAddr(s string) string {
+	idx := strings.LastIndex(s, ":")
+	if idx == -1 {
+		return s
+	}
+	return s[:idx]
+}
```

Comment:
I believe this will choke on IPv6 addresses, instead there is https://pkg.go.dev/net/netip#ParseAddrPort which should do the trick